metldata/load/load.py

```diff
@@ -48,10 +51,80 @@ async def load_artifacts_using_dao(
 ) -> None:
     """Load artifact resources from multiple submissions using the given dao collection."""
 
+    (
+        removed_resource_tags,
+        new_resources,
+        changed_resources,
+    ) = await _get_changed_resources(
+        artifact_resources=artifact_resources,
+        artifact_info_dict=artifact_info_dict,
+        dao_collection=dao_collection,
+    )
+
+    await process_removed_resources(
+        resource_tags=removed_resource_tags, dao_collection=dao_collection
+    )
+
+    await process_new_or_changed_resources(
+        resources=new_resources, dao_collection=dao_collection
+    )
+
+    await process_new_or_changed_resources(
+        resources=changed_resources, dao_collection=dao_collection
+    )
+
+
+async def _get_changed_resources(  # pylint: disable=too-many-locals
+    artifact_resources: ArtifactResourceDict,
+    artifact_info_dict: dict[str, ArtifactInfo],
+    dao_collection: ArtifactDaoCollection,
+) -> tuple[
+    set[tuple[str, str, str]],
+    dict[tuple[str, str, str], ArtifactResource],
+    dict[tuple[str, str, str], ArtifactResource],
+]:
+    """Extract changed resources using DAOs and currently submitted artifacts
+
+    Returns the following collections in order:
+     - A set of resource tags for artifact resources that have been removed
+     - A dict of new artifact resources indexed by the corresponding resource_tag
+     - A dict of changed artifact resources indexed by the corresponding resource_tag
+    """
+
+    # only need to collect resource tag to check if something needs to be deleted
+    unchanged_resource_tags = []
+
+    # collect new/changed resources indexed by their resource_tag for insert/upsert
+    new_resources = {}
+    changed_resources = {}
+
+    existing_resource_tags = await dao_collection.get_all_resource_tags()
```

Comment:
You can push that line down now after the nested loop.

Reply:
Still need that for one for the new_resource check in the loop

Reply:
Ah right. Then it might be good to convert it to a set here instead of doing that later.

Reply:
Maybe you should also reverse the two nested if/else by using "=" instead of "!=" and "in" instead of "not in". That's a bit easier to read.